[STORE] Add simple cache for StoreStats

s1monw opened the pull request:
this commit tries to reduce the filesystem calls to fetch metadata
by using a simple cache on top of the stats call.

Relates to #9683

kimchy commented:
This looks great!, a few comments:

- You changed `OsService` to use the new refresh cache thingy, we should also do it on `FsService`, `NetworkService`, and `ProcessService`.

- The hasChanged flag, if the dir changes not by creating an output, like in the context of shadow replica, then it won't be tripped to recalculate, right? What should we do in this case? I think that maybe we should not use hasChanged, but still maintain a cache of known files and their length, that are created through createOutput (and use it only in estimateSize) so we reduce the length metadata calls?

s1monw commented:
> You changed OsService to use the new refresh cache thingy, we should also do it on FsService, NetworkService, and ProcessService.

sure will do


> The hasChanged flag, if the dir changes not by creating an output, like in the context of shadow replica, then it won't be tripped to recalculate, right? What should we do in this case? I think that maybe we should not use hasChanged, but still maintain a cache of known files and their length, that are created through createOutput (and use it only in estimateSize) so we reduce the length metadata calls

the more I think about this the more I think we should try a different way. I wonder if we can use the Java 7 [WatchService](http://docs.oracle.com/javase/7/docs/api/java/nio/file/WatchService.html) and register something on the data dir to get notifications if something has changed from the OS instead of this crazy caching here?

bleskes commented on src/main/java/org/elasticsearch/common/util/SingleObjectCache.java:
```diff
@@ -0,0 +1,73 @@
+        if (needsRefresh()) {
+            if(refreshLock.tryLock()) {
+                try {
+                    cached = refresh();
```
do we need another needsRefresh check here?

kimchy commented:
> the more I think about this the more I think we should try a different way. I wonder if we can use the Java 7 WatchService and register something on the data dir to get notifications if something has changed from the OS instead of this crazy caching here?

I like this idea!

bleskes commented:
left one minor comment.  +1 on listening to file system changes.

s1monw commented:
@kimchy @bleskes I moved back and went with the cache only solution for now.  I added the new cache to all the other places and fixed some tests. can you review?

bleskes commented on src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -114,6 +119,11 @@ public Store(ShardId shardId, @IndexSettings Settings indexSettings, DirectorySe
         this.onClose = onClose;
+        final TimeValue refreshInterval = indexSettings.getAsTime(INDEX_STORE_STATS_REFRESH_INTERVAL, TimeValue.timeValueSeconds(10));
+        this.statsCache = new StoreStatsCache(refreshInterval);
+        statsCache.getOrRefresh();
```
do we need to do this on the constructor?

bleskes replied:
never mind - I see now we always rely on a valid cached object (because of the non blocking nature). In this case I wonder if we should force passing it to the constructor?

bleskes commented on src/main/java/org/elasticsearch/monitor/network/NetworkService.java:
```diff
@@ -37,23 +40,20 @@
         this.info.refreshInterval = refreshInterval.millis();
-        this.cachedStats = probe.networkStats();
-
+        networkStatsCache = new NetworkStatsCache(refreshInterval);
```
I think we need a networkStatsCache.getOrRefresh()?

bleskes commented on src/main/java/org/elasticsearch/monitor/process/ProcessService.java:
```diff
@@ -30,24 +32,18 @@
-        this.refreshInterval = componentSettings.getAsTime("refresh_interval", TimeValue.timeValueSeconds(1));
-
+        final TimeValue refreshInterval = settings.getAsTime("monitor.process.refresh_interval", TimeValue.timeValueSeconds(1));
+        processStatsCache = new ProcessStatsCache(refreshInterval);
```
same here - we miss a processStatsCache.getOrRefresh?

bleskes commented on src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryTests.java:
```diff
@@ -401,7 +402,8 @@ private IndicesStatsResponse createAndPopulateIndex(String name, int nodeCount,
         assertAcked(prepareCreate(name, nodeCount, settingsBuilder().put("number_of_shards", shardCount)
-                .put("number_of_replicas", replicaCount)));
+                .put("number_of_replicas", replicaCount)
+        .put(Store.INDEX_STORE_STATS_REFRESH_INTERVAL, 0)));
```
minor formatting issue

bleskes commented:
Looks much cleaner. I left some comments w.r.t the intialization logic of the cache, where the current semantics require the cached object to be non-null. Maybe we should lock in that case (instead of tryLock)

s1monw commented:
@bleskes pushed a new commit - this requires a non-null obj in the ctor.

bleskes commented:
+1
